fn.go

```diff
@@ -58,148 +56,115 @@ func (f *Function) RunFunction(ctx context.Context, req *fnv1beta1.RunFunctionRe
-		return rsp, nil
+	// TODO(lcarelli): handle early return when desired is empty. If you spot this on review, please mention it.
 
-	}
-
-	tagFilters, err := resolveTagFilters(in, xr, desiredComposed)
-	if err != nil {
-		f.log.Info("Failed to resolve tag filters.",
-			"error", err,
-			"tagFilters", in.TagFilters,
-			"xr", xr,
-			"managedResource", desiredComposed,
+	if resources.LenObserved() > 0 && resources.AllHaveExternalNamesSet() {
+		externalNames := resources.ObservedExternalNames()
+		f.log.Debug("External name already set for all resources",
+			"externalNames", externalNames,
 		)
-		response.Fatal(rsp, errors.Wrapf(err, "cannot resolve tag filters"))
+		response.Normalf(rsp, "external name annotation already set for all resources: %v", externalNames)
 		return rsp, nil
 	}
 
-	paginator := resourcegroupstaggingapi.NewGetResourcesPaginator(f.client, &resourcegroupstaggingapi.GetResourcesInput{
-		TagFilters: tagFilters,
-	})
+	err = resources.ForEachDesiredComposed(func(desiredComposed internal.Resource) error {
+		xr, err := request.GetObservedCompositeResource(req)
+		if err != nil {
+			return fmt.Errorf("extracting observed XR from req: %v", err)
+		}
 
-	var tagMappings []types.ResourceTagMapping
-	for paginator.HasMorePages() {
-		page, err := paginator.NextPage(context.Background())
+		tagFilters, err := resolveTagFilters(in, xr, desiredComposed)
 		if err != nil {
-			f.log.Info("Failed to paginate resource tag mappings.",
+			f.log.Info("Failed to resolve tag filters.",
 				"error", err,
+				"tagFilters", in.TagFilters,
+				"xr", xr,
+				"managedResource", desiredComposed,
 			)
-			response.Fatal(rsp, errors.Wrapf(err, "cannot get resources tag mappings"))
-			return rsp, nil
+			return fmt.Errorf("resolving tag filters: %v", err)
 		}
 
-		for _, t := range page.ResourceTagMappingList {
-			tagMappings = append(tagMappings, t)
+		paginator := resourcegroupstaggingapi.NewGetResourcesPaginator(f.client, &resourcegroupstaggingapi.GetResourcesInput{
+			TagFilters: tagFilters,
+		})
+
+		var tagMappings []types.ResourceTagMapping
+		for paginator.HasMorePages() {
+			page, err := paginator.NextPage(context.Background())
+			if err != nil {
+				return fmt.Errorf("getting resources tag mappings: %v", err)
+			}
+
+			for _, t := range page.ResourceTagMappingList {
+				tagMappings = append(tagMappings, t)
+			}
 		}
-	}
 
-	if len(tagMappings) > 1 {
-		f.log.Info("Ambiguous tag filters.", // TODO(lcaparelli): better word than ambiguous, maybe better message overall
-			"error", errors.New("found more than one resource matching tag filters"),
-			"tagFilters", tagFilters,
-			"matchingResources", extractARNs(tagMappings),
-		)
-		response.Fatal(rsp, fmt.Errorf("found more than one resource matching tag filters: %v", extractARNs(tagMappings)))
-		return rsp, nil
-	}
+		if len(tagMappings) > 1 {
+			f.log.Info("Ambiguous tag filters.", // TODO(lcaparelli): better word than ambiguous, maybe better message overall
```

Comment:
Maybe nonunique?

Reply:
The result is non-unique, but not the tag filters themselves... :thinking: 

Also thought about "Unresolvable tag filters", but they are resolvable. In fact more than once, hence the problem lol

Maybe "Cannot decide which resource to import."?

Reply:
> Maybe "Cannot decide which resource to import."?

Running with that one, let me know if you can come up with a better one and we can adjust it :grin: